plugins/hls-overloaded-record-dot-plugin/src/Ide/Plugin/OverloadedRecordDot.hs

```diff
@@ -0,0 +1,244 @@
+                        location     :: Range,
+                        -- | the record selector, this is found in front of recordExpr, but
+                        -- | get's placed after it when converted into record dot syntax
+                        selectorExpr :: LHsExpr (GhcPass 'Renamed),
+                        -- | the record. Whereas it can be many different structures, the only
+                        -- | requirement is that it evaluates to a record in the end
+                        recordExpr   :: LHsExpr (GhcPass 'Renamed) }
+
+instance Pretty RecordSelectorExpr where
+  pretty (RecordSelectorExpr l rs se) = pretty (printOutputable rs) <> ":" <+> pretty (printOutputable se)
+
+instance NFData RecordSelectorExpr where
+  rnf = rwhnf
+
+descriptor :: Recorder (WithPriority Log) -> PluginId -> PluginDescriptor IdeState
+descriptor recorder plId = (defaultPluginDescriptor plId)
+  { pluginHandlers = mkPluginHandler STextDocumentCodeAction codeActionProvider
+  , pluginRules = collectRecSelsRule recorder
+  }
+
+codeActionProvider :: PluginMethodHandler IdeState 'TextDocumentCodeAction
+codeActionProvider ideState pId (CodeActionParams _ _ caDocId caRange _) = pluginResponse $ do
+  nfp <- getNormalizedFilePath (caDocId ^. L.uri)
+  pragma <- getFirstPragma pId ideState nfp
+  CRSR crsMap exts <- collectRecSelResult ideState nfp
+  let
+    pragmaEdit = if OverloadedRecordDot `elem` exts
+                            then Nothing
+                            else Just $ insertNewPragma pragma OverloadedRecordDot
+    edits crs = convertRecordSelectors crs : maybeToList pragmaEdit
+    changes crs = Just $ HashMap.singleton (fromNormalizedUri (normalizedFilePathToUri nfp)) (List (edits crs))
+    mkCodeAction crs = InR CodeAction
+        { -- we pass the record selector to the title function, so that we can have the name of the record selector
+          -- in the title of the codeAction allowing the user can easily distinguish between the different codeActions
+          -- when using nested record selectors, the disadvantage is we need to print out the name of the record selector
+          -- which will decrease performance
+          _title = mkCodeActionTitle exts crs
+        , _kind = Just CodeActionRefactorRewrite
+        , _diagnostics = Nothing
+        , _isPreferred = Nothing
+        , _disabled = Nothing
+        , _edit = Just $ WorkspaceEdit (changes crs) Nothing Nothing
+        , _command = Nothing
+        , _xdata = Nothing
+        }
+    actions = map mkCodeAction (RangeMap.filterByRange caRange crsMap)
+  pure $ List actions
+  where
+    mkCodeActionTitle :: [Extension] -> RecordSelectorExpr-> Text
+    mkCodeActionTitle exts (RecordSelectorExpr _ se _) =
+      if OverloadedRecordDot `elem` exts
+        then title
+        else title <> " (needs extension: OverloadedRecordDot)"
+        where
+          title = "Convert `" <> name <> "` to record dot syntax"
```

Comment:
Just thinking about how this will read. If I understand right, for something like
```
foo bar
```
we'll get "Convert `foo` to record dot syntax". This reads a _little_ weirdly to me, since "foo" is not really the thing being converted? What about something like "Use record dot syntax to access `foo`"?

Reply:
There is not really a limit to how long the "record" will be, because we only require it to be an expression that eventually evaluates to a record. The record selector, on the other hand, should be relatively short. I like the "Convert `foo` to record syntax" option, because, that's the part that's actually changing. I feel "Use record dot syntax to access `foo`" would be incorrect, as we are not accessing foo, rather we are using foo to access bar. (foo as a record selector, is effectively a function that allows us to acesss part of bar)